burger/toppings/pluginchannels.py

```diff
@@ -0,0 +1,142 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from .topping import Topping
+
+from jawa.constants import String
+from jawa.classloader import ClassLoader
+
+import re
+
+_CHANNEL_IDENTIFIER = re.compile("^(minecraft:)?[a-z0-9/_.]+$")
+_CHANNEL_STRING = re.compile("^MC\|[a-zA-Z0-9]+$")
+
+class PluginChannelsTopping(Topping):
+    """Provides a list of all plugin channels"""
+
+    PROVIDES = ["pluginchannels"]
+    DEPENDS = [
+        "identify.nethandler.client",
+        "version.id",
+        "version.protocol"
+    ]
+
+    @staticmethod
+    def act(aggregate, classloader, verbose=False):
+        _check_integrity(aggregate)
+
+        pluginchannels = aggregate.setdefault("pluginchannels", {})
+        post_netty, protocol = _get_version_info(aggregate)
+
+        if not post_netty and protocol < 39:
+            # No plugin channels (likely) existed before 12w30c
+            return
+
+        if post_netty and protocol >= 385:
+            # After and during 1.13-pre3, the channels are identifiers declared in the two custom payload packet classes
+            # The internal channels use the format "minecraft:<channel>"
+            packets = _get_custom_payload_packets(classloader)
+            all_channels = [_get_class_constants(classloader, side, _is_channel_identifier) for side in packets]
+
+            if protocol >= 443:
+                # After 18w43c, channels are not explicitly defined with the "minecraft" namespace
+                # Semanticas don't change, so let's add it back for consistency between versions
+                all_channels = [[f"minecraft:{channel}" for channel in channels] for channels in all_channels]
+        else:
+            # Before and during 1.13-pre2, the channels are strings declared in the two play packet handlers
+            # The internal channels use the format "MC|<channel>"
+            nethandlers = _get_nethandlers(aggregate, classloader)
+            all_channels = [_get_class_constants(classloader, side, _is_channel_string) for side in nethandlers]
+
+        for channels in all_channels:
+                channels.sort()
+
+        pluginchannels["clientbound"], pluginchannels["serverbound"] = all_channels
+
+def _check_integrity(aggregate):
+    required_fields = ["classes", "version"]
+
+    for field in required_fields:
+        assert field in aggregate, f"{field} is missing from aggregate"
+
+def _get_version_info(aggregate):
+    id = aggregate["version"].get("id")
+    protocol = aggregate["version"].get("protocol")
+
+    assert id is not None, "version.id is missing from aggregate"
+    assert protocol is not None, "version.protocol is missing from aggregate"
+    
+    return (id != "", protocol)
```

Comment:
I'm surprised this works; it seems a bit counterintuitive that `id` would be present but empty, but it seems to be the case that (presumably) 13w39b and earlier use string concatenation (e.g. `"Outdated server! I'm still on " + version`) while 13w41a and later have the string inlined (e.g. `"Outdated server! I'm still on 13w41a"). This should probably be mentioned in a comment.

This technique doesn't work further back; [1.1 added plugin channels](https://web.archive.org/web/20120115204533/http://dinnerbone.com/blog/2012/01/13/minecraft-plugin-channels-messaging), but it seems like the "Outdated server! I'm still on" check only exists starting in 12w18a (and `id` isn't set in 12w17a and earlier (including 1.2.5) but is present in 1.3.1 and later).

Reply:
Initially, I was checking the dependencies of the client nethandler for mentions of any netty-related class, and using that to define if the version was post netty. But then I found this pattern of the id conveniently being blank on pre netty versions for some reason (I didn't really look into as to why), and decided to stick with it instead since it was simpler (albeit less intuituve).

But if this pattern is untintentional (and may potentially break if the version topping changes), then I can just revert back to the old check system.

Reply:
It seems like a reasonable pattern, but it might be better to add a `netty_rewrite` field in the version topping instead to make it more obvious.

Reply:
That would be nice, I was going to suggest it